File: server/integrations/zero_g_storage.py
"""
0G Storage Integration for iNFT Memory Persistence
Provides secure upload/download of encrypted AI/agent memory to decentralized 0G Storage
"""

import os
import hashlib
import sqlite3
import asyncio
import logging
from typing import Dict, List, Optional, Tuple, Union
from pathlib import Path
from datetime import datetime
import json
import base64
from dataclasses import dataclass, asdict
from cryptography.fernet import Fernet
from cryptography.hazmat.primitives import hashes
from cryptography.hazmat.primitives.kdf.pbkdf2 import PBKDF2HMAC
from web3 import Web3
from web3.contract import Contract

logger = logging.getLogger(__name__)


@dataclass
class StorageMetadata:
    """Metadata for stored iNFT memory"""
    inft_id: str
    file_hash: str
    checksum: str
    timestamp: int
    size_bytes: int
    encryption_key_id: str
    version: int = 1


@dataclass
class EventLogBatch:
    """Batch of append-only event logs"""
    inft_id: str
    events: List[Dict]
    batch_id: str
    timestamp: int
    previous_hash: Optional[str] = None


class ZeroGStorageClient:
    """
    Client for interacting with 0G Storage network for iNFT memory persistence
    """
    
    def __init__(
        self,
        rpc_url: str,
        storage_contract_address: str,
        private_key: Optional[str] = None,
        encryption_key: Optional[bytes] = None
    ):
        """
        Initialize 0G Storage client
        
        Args:
            rpc_url: 0G network RPC endpoint
            storage_contract_address: Storage registry contract address
            private_key: Optional private key for signing transactions
            encryption_key: Optional key for encrypting/decrypting data
        """
        self.w3 = Web3(Web3.HTTPProvider(rpc_url))
        self.storage_address = Web3.to_checksum_address(storage_contract_address)
        
        if private_key:
            self.account = self.w3.eth.account.from_key(private_key)
        else:
            self.account = None
        
        # Initialize encryption
        self.encryption_key = encryption_key
        if encryption_key:
            self.cipher = Fernet(encryption_key)
        else:
            self.cipher = None
        
        # Storage contract ABI (minimal interface)
        self.storage_abi = [
            {
                "inputs": [
                    {"name": "inftId", "type": "string"},
                    {"name": "storageHash", "type": "string"},
                    {"name": "checksum", "type": "string"}
                ],
                "name": "updateStoragePointer",
                "outputs": [],
                "stateMutability": "nonpayable",
                "type": "function"
            },
            {
                "inputs": [{"name": "inftId", "type": "string"}],
                "name": "getStoragePointer",
                "outputs": [
                    {"name": "storageHash", "type": "string"},
                    {"name": "checksum", "type": "string"},
                    {"name": "timestamp", "type": "uint256"}
                ],
                "stateMutability": "view",
                "type": "function"
            }
        ]
        
        self.storage_contract = self.w3.eth.contract(
            address=self.storage_address,
            abi=self.storage_abi
        )
    
    def calculate_file_checksum(self, file_path: str) -> str:
        """
        Calculate SHA-256 checksum of a file
        
        Args:
            file_path: Path to file
        
        Returns:
            Hex-encoded SHA-256 checksum
        """
        sha256_hash = hashlib.sha256()
        with open(file_path, "rb") as f:
            for byte_block in iter(lambda: f.read(4096), b""):
                sha256_hash.update(byte_block)
        return sha256_hash.hexdigest()
    
    def encrypt_file(self, input_path: str, output_path: str) -> str:
        """
        Encrypt a file using Fernet symmetric encryption
        
        Args:
            input_path: Path to input file
            output_path: Path to encrypted output file
        
        Returns:
            Checksum of encrypted file
        
        Raises:
            ValueError: If encryption key is not configured
        """
        if not self.cipher:
            raise ValueError("Encryption key not configured")
        
        with open(input_path, "rb") as f:
            data = f.read()
        
        encrypted_data = self.cipher.encrypt(data)
        
        with open(output_path, "wb") as f:
            f.write(encrypted_data)
        
        return self.calculate_file_checksum(output_path)
    
    def decrypt_file(self, input_path: str, output_path: str) -> str:
        """
        Decrypt a file using Fernet symmetric encryption
        
        Args:
            input_path: Path to encrypted file
            output_path: Path to decrypted output file
        
        Returns:
            Checksum of decrypted file
        
        Raises:
            ValueError: If encryption key is not configured
        """
        if not self.cipher:
            raise ValueError("Encryption key not configured")
        
        with open(input_path, "rb") as f:
            encrypted_data = f.read()
        
        decrypted_data = self.cipher.decrypt(encrypted_data)
        
        with open(output_path, "wb") as f:
            f.write(decrypted_data)
        
        return self.calculate_file_checksum(output_path)
    
    async def upload_to_0g_storage(self, file_path: str) -> Tuple[str, int]:
        """
        Upload file to 0G Storage network
        
        Args:
            file_path: Path to file to upload
        
        Returns:
            Tuple of (storage_hash, file_size)
        
        Note:
            This is a placeholder implementation. In production, this would
            interact with the actual 0G Storage SDK/API to upload the file
            and return the content-addressed hash.
        """
        # Calculate file hash as content identifier
        file_hash = self.calculate_file_checksum(file_path)
        file_size = os.path.getsize(file_path)
        
        # TODO: Implement actual 0G Storage upload using SDK
        # Example: storage_hash = await og_client.upload_file(file_path)
        
        logger.info(f"Uploaded file to 0G Storage: {file_hash} ({file_size} bytes)")
        
        # Return content hash and size
        return file_hash, file_size
    
    async def download_from_0g_storage(self, storage_hash: str, output_path: str) -> int:
        """
        Download file from 0G Storage network
        
        Args:
            storage_hash: Content hash of file in 0G Storage
            output_path: Local path to save downloaded file
        
        Returns:
            Size of downloaded file in bytes
        
        Note:
            This is a placeholder implementation. In production, this would
            interact with the actual 0G Storage SDK/API to download the file.
        """
        # TODO: Implement actual 0G Storage download using SDK
        # Example: await og_client.download_file(storage_hash, output_path)
        
        # Verify file integrity after download
        if os.path.exists(output_path):
            downloaded_hash = self.calculate_file_checksum(output_path)
            if downloaded_hash != storage_hash:
                logger.warning(f"Checksum mismatch: expected {storage_hash}, got {downloaded_hash}")
        
        file_size = os.path.getsize(output_path) if os.path.exists(output_path) else 0
        logger.info(f"Downloaded file from 0G Storage: {storage_hash} ({file_size} bytes)")
        
        return file_size
    
    async def update_inft_storage_pointer(
        self,
        inft_id: str,
        storage_hash: str,
        checksum: str
    ) -> Dict:
        """
        Update on-chain storage pointer for an iNFT
        
        Args:
            inft_id: Unique identifier for the iNFT
            storage_hash: 0G Storage content hash
            checksum: File integrity checksum
        
        Returns:
            Transaction receipt
        
        Raises:
            ValueError: If account is not configured
        """
        if not self.account:
            raise ValueError("Private key required for transactions")
        
        # Build transaction
        tx = self.storage_contract.functions.updateStoragePointer(
            inft_id,
            storage_hash,
            checksum
        ).build_transaction({
            'from': self.account.address,
            'gas': 150000,
            'gasPrice': self.w3.eth.gas_price,
            'nonce': self.w3.eth.get_transaction_count(self.account.address),
        })
        
        # Sign and send
        signed_tx = self.w3.eth.account.sign_transaction(tx, self.account.key)
        tx_hash = self.w3.eth.send_raw_transaction(signed_tx.rawTransaction)
        
        # Wait for confirmation
        receipt = self.w3.eth.wait_for_transaction_receipt(tx_hash)
        
        return {
            'tx_hash': tx_hash.hex(),
            'status': receipt['status'],
            'gas_used': receipt['gasUsed'],
            'block_number': receipt['blockNumber']
        }
    
    async def get_inft_storage_pointer(self, inft_id: str) -> Tuple[str, str, int]:
        """
        Get current storage pointer for an iNFT
        
        Args:
            inft_id: Unique identifier for the iNFT
        
        Returns:
            Tuple of (storage_hash, checksum, timestamp)
        """
        result = self.storage_contract.functions.getStoragePointer(inft_id).call()
        return result
    
    async def sync_to_0g_storage(
        self,
        inft_id: str,
        db_path: str,
        encrypt: bool = True
    ) -> StorageMetadata:
        """
        Complete workflow: encrypt, upload SQLite DB to 0G Storage, update on-chain pointer
        
        Args:
            inft_id: Unique identifier for the iNFT
            db_path: Path to SQLite database file
            encrypt: Whether to encrypt before upload (default: True)
        
        Returns:
            StorageMetadata with upload details
        
        Raises:
            FileNotFoundError: If database file doesn't exist
            ValueError: If encryption is requested but key not configured
        """
        if not os.path.exists(db_path):
            raise FileNotFoundError(f"Database file not found: {db_path}")
        
        # Validate database integrity
        try:
            conn = sqlite3.connect(db_path)
            conn.execute("PRAGMA integrity_check")
            conn.close()
        except Exception as e:
            raise ValueError(f"Database integrity check failed: {e}")
        
        # Encrypt if requested
        upload_path = db_path
        if encrypt:
            if not self.cipher:
                raise ValueError("Encryption requested but key not configured")
            
            encrypted_path = f"{db_path}.encrypted"
            self.encrypt_file(db_path, encrypted_path)
            upload_path = encrypted_path
        
        # Calculate checksum before upload
        checksum = self.calculate_file_checksum(upload_path)
        
        # Upload to 0G Storage
        storage_hash, file_size = await self.upload_to_0g_storage(upload_path)
        
        # Update on-chain pointer
        if self.account:
            tx_result = await self.update_inft_storage_pointer(
                inft_id,
                storage_hash,
                checksum
            )
            logger.info(f"Updated on-chain pointer: {tx_result['tx_hash']}")
        
        # Clean up encrypted file if created
        if encrypt and os.path.exists(encrypted_path):
            os.remove(encrypted_path)
        
        # Create metadata
        metadata = StorageMetadata(
            inft_id=inft_id,
            file_hash=storage_hash,
            checksum=checksum,
            timestamp=int(datetime.now().timestamp()),
            size_bytes=file_size,
            encryption_key_id="default" if encrypt else "none"
        )
        
        logger.info(f"Synced iNFT {inft_id} to 0G Storage: {storage_hash}")
        
        return metadata
    
    async def load_from_0g_storage(
        self,
        inft_id: str,
        output_dir: str = "/tmp",
        verify_checksum: bool = True
    ) -> str:
        """
        Complete workflow: download database from 0G Storage, decrypt, validate
        
        Args:
            inft_id: Unique identifier for the iNFT
            output_dir: Directory to save downloaded database
            verify_checksum: Whether to verify checksum after download
        
        Returns:
            Path to restored database file
        
        Raises:
            ValueError: If checksum verification fails
        """
        # Get storage pointer from chain
        storage_hash, expected_checksum, timestamp = await self.get_inft_storage_pointer(inft_id)
        
        logger.info(f"Loading iNFT {inft_id} from 0G Storage: {storage_hash}")
        
        # Download from 0G Storage
        os.makedirs(output_dir, exist_ok=True)
        encrypted_path = os.path.join(output_dir, f"inft_{inft_id}.db.encrypted")
        
        await self.download_from_0g_storage(storage_hash, encrypted_path)
        
        # Verify checksum
        if verify_checksum:
            actual_checksum = self.calculate_file_checksum(encrypted_path)
            if actual_checksum != expected_checksum:
                raise ValueError(
                    f"Checksum verification failed: expected {expected_checksum}, "
                    f"got {actual_checksum}"
                )
        
        # Decrypt if encrypted
        final_path = os.path.join(output_dir, f"inft_{inft_id}.db")
        
        if self.cipher:
            try:
                self.decrypt_file(encrypted_path, final_path)
                os.remove(encrypted_path)
            except Exception as e:
                logger.error(f"Decryption failed: {e}")
                # If decryption fails, file might not be encrypted
                os.rename(encrypted_path, final_path)
        else:
            os.rename(encrypted_path, final_path)
        
        # Validate database integrity
        try:
            conn = sqlite3.connect(final_path)
            conn.execute("PRAGMA integrity_check")
            conn.close()
        except Exception as e:
            raise ValueError(f"Downloaded database integrity check failed: {e}")
        
        logger.info(f"Loaded iNFT {inft_id} database to: {final_path}")
        
        return final_path
    
    async def append_event_log(
        self,
        inft_id: str,
        event: Dict,
        auto_batch: bool = True,
        batch_size: int = 100
    ) -> Optional[str]:
        """
        Append event to incremental event log with optional auto-batching
        
        Args:
            inft_id: Unique identifier for the iNFT
            event: Event data to append
            auto_batch: Whether to automatically upload when batch size reached
            batch_size: Number of events to batch before auto-upload
        
        Returns:
            Storage hash if batch was uploaded, None otherwise
        """
        # Initialize or load event log
        log_path = f"/tmp/inft_{inft_id}_events.jsonl"
        
        # Append event with timestamp
        event_entry = {
            "timestamp": datetime.now().isoformat(),
            "event": event
        }
        
        with open(log_path, "a") as f:
            f.write(json.dumps(event_entry) + "\n")
        
        # Check if auto-batch threshold reached
        if auto_batch:
            with open(log_path, "r") as f:
                event_count = sum(1 for _ in f)
            
            if event_count >= batch_size:
                # Upload batch
                storage_hash, _ = await self.upload_to_0g_storage(log_path)
                
                # Create new log for next batch
                archive_path = f"/tmp/inft_{inft_id}_events_{int(datetime.now().timestamp())}.jsonl"
                os.rename(log_path, archive_path)
                
                logger.info(f"Auto-uploaded event batch for iNFT {inft_id}: {storage_hash}")
                
                return storage_hash
        
        return None
    
    def validate_chain_of_custody(
        self,
        inft_id: str,
        storage_hashes: List[str]
    ) -> bool:
        """
        Validate chain-of-custody for sequential storage updates
        
        Args:
            inft_id: Unique identifier for the iNFT
            storage_hashes: List of storage hashes in chronological order
        
        Returns:
            True if chain of custody is valid
        
        Note:
            This validates that each storage update builds upon the previous
            by checking hash linkage and temporal ordering
        """
        if len(storage_hashes) < 2:
            return True  # Single or no updates are trivially valid
        
        # Verify each hash links to previous
        for i in range(1, len(storage_hashes)):
            prev_hash = storage_hashes[i - 1]
            curr_hash = storage_hashes[i]
            
            # TODO: Implement actual chain validation logic
            # This would check that curr_hash includes prev_hash in its metadata
            # and that timestamps are monotonically increasing
            
            logger.debug(f"Validating custody: {prev_hash} -> {curr_hash}")
        
        return True
    
    def rotate_encryption_key(self, new_key: bytes) -> None:
        """
        Rotate encryption key for future operations
        
        Args:
            new_key: New encryption key
        
        Note:
            This only affects future encrypt/decrypt operations.
            Existing encrypted files must be re-encrypted manually.
        """
        self.encryption_key = new_key
        self.cipher = Fernet(new_key)
        logger.info("Encryption key rotated successfully")


# Convenience functions

def generate_encryption_key(password: str, salt: Optional[bytes] = None) -> bytes:
    """
    Generate encryption key from password using PBKDF2
    
    Args:
        password: User password
        salt: Optional salt (generated if not provided)
    
    Returns:
        Base64-encoded Fernet-compatible key
    """
    if salt is None:
        salt = os.urandom(16)
    
    kdf = PBKDF2HMAC(
        algorithm=hashes.SHA256(),
        length=32,
        salt=salt,
        iterations=100000,
    )
    
    key = base64.urlsafe_b64encode(kdf.derive(password.encode()))
    return key


async def sync_to_0g_storage(
    inft_id: str,
    db_path: str = None,
    storage_client: Optional[ZeroGStorageClient] = None
) -> StorageMetadata:
    """
    High-level convenience function to sync iNFT memory to 0G Storage
    
    Args:
        inft_id: Unique identifier for the iNFT
        db_path: Path to database file (default: /data/inft_{inft_id}.db)
        storage_client: Optional pre-configured client
    
    Returns:
        StorageMetadata with upload details
    """
    if db_path is None:
        db_path = f"/data/inft_{inft_id}.db"
    
    if storage_client is None:
        # Create client from environment variables
        rpc_url = os.getenv("ZERO_G_RPC", "https://evmrpc.0g.ai")
        storage_address = os.getenv("ZERO_G_STORAGE_CONTRACT")
        private_key = os.getenv("ZERO_G_PRIVATE_KEY")
        
        if not storage_address:
            raise ValueError("ZERO_G_STORAGE_CONTRACT environment variable required")
        
        # Generate or load encryption key
        encryption_password = os.getenv("INFT_ENCRYPTION_PASSWORD")
        encryption_key = None
        if encryption_password:
            encryption_key = generate_encryption_key(encryption_password)
        
        storage_client = ZeroGStorageClient(
            rpc_url=rpc_url,
            storage_contract_address=storage_address,
            private_key=private_key,
            encryption_key=encryption_key
        )
    
    return await storage_client.sync_to_0g_storage(inft_id, db_path)


async def load_from_0g_storage(
    inft_id: str,
    output_dir: str = "/data",
    storage_client: Optional[ZeroGStorageClient] = None
) -> str:
    """
    High-level convenience function to load iNFT memory from 0G Storage
    
    Args:
        inft_id: Unique identifier for the iNFT
        output_dir: Directory to save restored database
        storage_client: Optional pre-configured client
    
    Returns:
        Path to restored database file
    """
    if storage_client is None:
        # Create client from environment variables
        rpc_url = os.getenv("ZERO_G_RPC", "https://evmrpc.0g.ai")
        storage_address = os.getenv("ZERO_G_STORAGE_CONTRACT")
        private_key = os.getenv("ZERO_G_PRIVATE_KEY")
        
        if not storage_address:
            raise ValueError("ZERO_G_STORAGE_CONTRACT environment variable required")
        
        # Generate or load encryption key
        encryption_password = os.getenv("INFT_ENCRYPTION_PASSWORD")
        encryption_key = None
        if encryption_password:
            encryption_key = generate_encryption_key(encryption_password)
        
        storage_client = ZeroGStorageClient(
            rpc_url=rpc_url,
            storage_contract_address=storage_address,
            private_key=private_key,
            encryption_key=encryption_key
        )
    
    return await storage_client.load_from_0g_storage(inft_id, output_dir)


__all__ = [
    "ZeroGStorageClient",
    "StorageMetadata",
    "EventLogBatch",
    "generate_encryption_key",
    "sync_to_0g_storage",
    "load_from_0g_storage",
]

File: core/inft/zero-g-storage.ts
/**
 * 0G Storage Integration for iNFT Memory Persistence (TypeScript)
 * Provides secure upload/download of encrypted AI/agent memory to decentralized 0G Storage
 */

import * as crypto from 'crypto';
import * as fs from 'fs/promises';
import * as path from 'path';
import * as os from 'os';
import { ethers } from 'ethers';

/**
 * Metadata for stored iNFT memory
 */
export interface StorageMetadata {
  inftId: string;
  fileHash: string;
  checksum: string;
  timestamp: number;
  sizeBytes: number;
  encryptionKeyId: string;
  version: number;
}

/**
 * Batch of append-only event logs
 */
export interface EventLogBatch {
  inftId: string;
  events: Array<Record<string, any>>;
  batchId: string;
  timestamp: number;
  previousHash?: string;
}

/**
 * Configuration for ZeroGStorageClient
 */
export interface ZeroGStorageConfig {
  rpcUrl: string;
  storageContractAddress: string;
  privateKey?: string;
  encryptionKey?: Buffer;
}

/**
 * Storage pointer on-chain
 */
export interface StoragePointer {
  storageHash: string;
  checksum: string;
  timestamp: number;
}

/**
 * Client for interacting with 0G Storage network for iNFT memory persistence
 */
export class ZeroGStorageClient {
  private provider: ethers.JsonRpcProvider;
  private storageContract: ethers.Contract;
  private wallet?: ethers.Wallet;
  private encryptionKey?: Buffer;

  // Storage contract ABI (minimal interface)
  private static readonly STORAGE_ABI = [
    {
      inputs: [
        { name: 'inftId', type: 'string' },
        { name: 'storageHash', type: 'string' },
        { name: 'checksum', type: 'string' },
      ],
      name: 'updateStoragePointer',
      outputs: [],
      stateMutability: 'nonpayable',
      type: 'function',
    },
    {
      inputs: [{ name: 'inftId', type: 'string' }],
      name: 'getStoragePointer',
      outputs: [
        { name: 'storageHash', type: 'string' },
        { name: 'checksum', type: 'string' },
        { name: 'timestamp', type: 'uint256' },
      ],
      stateMutability: 'view',
      type: 'function',
    },
  ];

  /**
   * Initialize 0G Storage client
   */
  constructor(config: ZeroGStorageConfig) {
    this.provider = new ethers.JsonRpcProvider(config.rpcUrl);
    this.storageContract = new ethers.Contract(
      config.storageContractAddress,
      ZeroGStorageClient.STORAGE_ABI,
      this.provider
    );

    if (config.privateKey) {
      this.wallet = new ethers.Wallet(config.privateKey, this.provider);
      this.storageContract = this.storageContract.connect(this.wallet) as ethers.Contract;
    }

    this.encryptionKey = config.encryptionKey;
  }

  /**
   * Calculate SHA-256 checksum of a file
   */
  async calculateFileChecksum(filePath: string): Promise<string> {
    const data = await fs.readFile(filePath);
    const hash = crypto.createHash('sha256');
    hash.update(data);
    return hash.digest('hex');
  }

  /**
   * Encrypt a file using AES-256-GCM
   */
  async encryptFile(inputPath: string, outputPath: string): Promise<string> {
    if (!this.encryptionKey) {
      throw new Error('Encryption key not configured');
    }

    const data = await fs.readFile(inputPath);

    // Generate random IV
    const iv = crypto.randomBytes(16);

    // Create cipher
    const cipher = crypto.createCipheriv('aes-256-gcm', this.encryptionKey, iv);

    // Encrypt data
    const encrypted = Buffer.concat([cipher.update(data), cipher.final()]);

    // Get auth tag
    const authTag = cipher.getAuthTag();

    // Combine IV + authTag + encrypted data
    const combined = Buffer.concat([iv, authTag, encrypted]);

    await fs.writeFile(outputPath, combined);

    return this.calculateFileChecksum(outputPath);
  }

  /**
   * Decrypt a file using AES-256-GCM
   */
  async decryptFile(inputPath: string, outputPath: string): Promise<string> {
    if (!this.encryptionKey) {
      throw new Error('Encryption key not configured');
    }

    const data = await fs.readFile(inputPath);

    // Extract IV, auth tag, and encrypted data
    const iv = data.subarray(0, 16);
    const authTag = data.subarray(16, 32);
    const encrypted = data.subarray(32);

    // Create decipher
    const decipher = crypto.createDecipheriv('aes-256-gcm', this.encryptionKey, iv);
    decipher.setAuthTag(authTag);

    // Decrypt data
    const decrypted = Buffer.concat([decipher.update(encrypted), decipher.final()]);

    await fs.writeFile(outputPath, decrypted);

    return this.calculateFileChecksum(outputPath);
  }

  /**
   * Upload file to 0G Storage network
   * 
   * Note: This is a placeholder implementation. In production, this would
   * interact with the actual 0G Storage SDK/API.
   */
  async uploadTo0gStorage(filePath: string): Promise<{ storageHash: string; fileSize: number }> {
    // Calculate file hash as content identifier
    const fileHash = await this.calculateFileChecksum(filePath);
    const stats = await fs.stat(filePath);
    const fileSize = stats.size;

    // TODO: Implement actual 0G Storage upload using SDK
    // Example: const storageHash = await ogClient.uploadFile(filePath);

    console.log(`Uploaded file to 0G Storage: ${fileHash} (${fileSize} bytes)`);

    return { storageHash: fileHash, fileSize };
  }

  /**
   * Download file from 0G Storage network
   * 
   * Note: This is a placeholder implementation. In production, this would
   * interact with the actual 0G Storage SDK/API.
   */
  async downloadFrom0gStorage(storageHash: string, outputPath: string): Promise<number> {
    // TODO: Implement actual 0G Storage download using SDK
    // Example: await ogClient.downloadFile(storageHash, outputPath);

    // Verify file integrity after download
    try {
      const downloadedHash = await this.calculateFileChecksum(outputPath);
      if (downloadedHash !== storageHash) {
        console.warn(
          `Checksum mismatch: expected ${storageHash}, got ${downloadedHash}`
        );
      }
    } catch (error) {
      console.error('Checksum verification failed:', error);
    }

    const stats = await fs.stat(outputPath);
    const fileSize = stats.size;

    console.log(`Downloaded file from 0G Storage: ${storageHash} (${fileSize} bytes)`);

    return fileSize;
  }

  /**
   * Update on-chain storage pointer for an iNFT
   */
  async updateInftStoragePointer(
    inftId: string,
    storageHash: string,
    checksum: string
  ): Promise<ethers.TransactionReceipt> {
    if (!this.wallet) {
      throw new Error('Private key required for transactions');
    }

    const tx = await this.storageContract.updateStoragePointer(
      inftId,
      storageHash,
      checksum
    );

    const receipt = await tx.wait();

    return receipt;
  }

  /**
   * Get current storage pointer for an iNFT
   */
  async getInftStoragePointer(inftId: string): Promise<StoragePointer> {
    const result = await this.storageContract.getStoragePointer(inftId);

    return {
      storageHash: result[0],
      checksum: result[1],
      timestamp: Number(result[2]),
    };
  }

  /**
   * Complete workflow: encrypt, upload SQLite DB to 0G Storage, update on-chain pointer
   */
  async syncTo0gStorage(
    inftId: string,
    dbPath: string,
    encrypt: boolean = true
  ): Promise<StorageMetadata> {
    // Verify file exists
    try {
      await fs.access(dbPath);
    } catch (error) {
      throw new Error(`Database file not found: ${dbPath}`);
    }

    // Encrypt if requested
    let uploadPath = dbPath;
    if (encrypt) {
      if (!this.encryptionKey) {
        throw new Error('Encryption requested but key not configured');
      }

      const encryptedPath = `${dbPath}.encrypted`;
      await this.encryptFile(dbPath, encryptedPath);
      uploadPath = encryptedPath;
    }

    // Calculate checksum before upload
    const checksum = await this.calculateFileChecksum(uploadPath);

    // Upload to 0G Storage
    const { storageHash, fileSize } = await this.uploadTo0gStorage(uploadPath);

    // Update on-chain pointer
    if (this.wallet) {
      const receipt = await this.updateInftStoragePointer(inftId, storageHash, checksum);
      console.log(`Updated on-chain pointer: ${receipt.hash}`);
    }

    // Clean up encrypted file if created
    if (encrypt) {
      try {
        await fs.unlink(`${dbPath}.encrypted`);
      } catch (error) {
        // Ignore cleanup errors
      }
    }

    // Create metadata
    const metadata: StorageMetadata = {
      inftId,
      fileHash: storageHash,
      checksum,
      timestamp: Math.floor(Date.now() / 1000), // Use seconds for consistency with Python
      sizeBytes: fileSize,
      encryptionKeyId: encrypt ? 'default' : 'none',
      version: 1,
    };

    console.log(`Synced iNFT ${inftId} to 0G Storage: ${storageHash}`);

    return metadata;
  }

  /**
   * Complete workflow: download database from 0G Storage, decrypt, validate
   */
  async loadFrom0gStorage(
    inftId: string,
    outputDir?: string,
    verifyChecksum: boolean = true
  ): Promise<string> {
    // Use system temp directory if not specified
    const finalOutputDir = outputDir || os.tmpdir();
    // Get storage pointer from chain
    const { storageHash, checksum: expectedChecksum } =
      await this.getInftStoragePointer(inftId);

    console.log(`Loading iNFT ${inftId} from 0G Storage: ${storageHash}`);

    // Download from 0G Storage
    await fs.mkdir(finalOutputDir, { recursive: true });
    const encryptedPath = path.join(finalOutputDir, `inft_${inftId}.db.encrypted`);

    await this.downloadFrom0gStorage(storageHash, encryptedPath);

    // Verify checksum
    if (verifyChecksum) {
      const actualChecksum = await this.calculateFileChecksum(encryptedPath);
      if (actualChecksum !== expectedChecksum) {
        throw new Error(
          `Checksum verification failed: expected ${expectedChecksum}, got ${actualChecksum}`
        );
      }
    }

    // Decrypt if encrypted
    const finalPath = path.join(finalOutputDir, `inft_${inftId}.db`);

    if (this.encryptionKey) {
      try {
        await this.decryptFile(encryptedPath, finalPath);
        await fs.unlink(encryptedPath);
      } catch (error) {
        console.error('Decryption failed:', error);
        // If decryption fails, file might not be encrypted
        await fs.rename(encryptedPath, finalPath);
      }
    } else {
      await fs.rename(encryptedPath, finalPath);
    }

    console.log(`Loaded iNFT ${inftId} database to: ${finalPath}`);

    return finalPath;
  }

  /**
   * Append event to incremental event log with optional auto-batching
   * 
   * Note: Archived log files (when batch uploads) are not automatically cleaned up.
   * Consider implementing a cleanup strategy based on your retention policies.
   */
  async appendEventLog(
    inftId: string,
    event: Record<string, any>,
    autoBatch: boolean = true,
    batchSize: number = 100,
    logDir?: string
  ): Promise<string | null> {
    // Use system temp directory if not specified
    const finalLogDir = logDir || os.tmpdir();
    
    // Initialize or load event log
    const logPath = path.join(finalLogDir, `inft_${inftId}_events.jsonl`);

    // Append event with timestamp
    const eventEntry = {
      timestamp: new Date().toISOString(),
      event,
    };

    await fs.appendFile(logPath, JSON.stringify(eventEntry) + '\n');

    // Check if auto-batch threshold reached
    if (autoBatch) {
      const content = await fs.readFile(logPath, 'utf-8');
      const eventCount = content.split('\n').filter((line) => line.trim()).length;

      if (eventCount >= batchSize) {
        // Upload batch
        const { storageHash } = await this.uploadTo0gStorage(logPath);

        // Archive old log for next batch (user should implement cleanup)
        // Uses milliseconds for archive filename (consistent with Python implementation)
        const archivePath = path.join(finalLogDir, `inft_${inftId}_events_${Date.now()}.jsonl`);
        await fs.rename(logPath, archivePath);

        console.log(`Auto-uploaded event batch for iNFT ${inftId}: ${storageHash}`);
        console.log(`Archived log to: ${archivePath} (cleanup not automatic)`);

        return storageHash;
      }
    }

    return null;
  }

  /**
   * Validate chain-of-custody for sequential storage updates
   */
  validateChainOfCustody(inftId: string, storageHashes: string[]): boolean {
    if (storageHashes.length < 2) {
      return true; // Single or no updates are trivially valid
    }

    // Verify each hash links to previous
    for (let i = 1; i < storageHashes.length; i++) {
      const prevHash = storageHashes[i - 1];
      const currHash = storageHashes[i];

      // TODO: Implement actual chain validation logic
      // This would check that currHash includes prevHash in its metadata
      // and that timestamps are monotonically increasing

      console.debug(`Validating custody: ${prevHash} -> ${currHash}`);
    }

    return true;
  }

  /**
   * Rotate encryption key for future operations
   */
  rotateEncryptionKey(newKey: Buffer): void {
    this.encryptionKey = newKey;
    console.log('Encryption key rotated successfully');
  }
}

/**
 * Generate encryption key from password using PBKDF2
 */
export function generateEncryptionKey(password: string, salt?: Buffer): Buffer {
  const actualSalt = salt || crypto.randomBytes(16);

  return crypto.pbkdf2Sync(password, actualSalt, 100000, 32, 'sha256');
}

/**
 * High-level convenience function to sync iNFT memory to 0G Storage
 */
export async function syncTo0gStorage(
  inftId: string,
  dbPath?: string,
  storageClient?: ZeroGStorageClient
): Promise<StorageMetadata> {
  const finalDbPath = dbPath || `/data/inft_${inftId}.db`;

  let client = storageClient;
  if (!client) {
    // Create client from environment variables
    const rpcUrl = process.env.ZERO_G_RPC || 'https://evmrpc.0g.ai';
    const storageAddress = process.env.ZERO_G_STORAGE_CONTRACT;
    const privateKey = process.env.ZERO_G_PRIVATE_KEY;

    if (!storageAddress) {
      throw new Error('ZERO_G_STORAGE_CONTRACT environment variable required');
    }

    // Generate or load encryption key
    const encryptionPassword = process.env.INFT_ENCRYPTION_PASSWORD;
    let encryptionKey: Buffer | undefined;
    if (encryptionPassword) {
      encryptionKey = generateEncryptionKey(encryptionPassword);
    }

    client = new ZeroGStorageClient({
      rpcUrl,
      storageContractAddress: storageAddress,
      privateKey,
      encryptionKey,
    });
  }

  return client.syncTo0gStorage(inftId, finalDbPath);
}

/**
 * High-level convenience function to load iNFT memory from 0G Storage
 */
export async function loadFrom0gStorage(
  inftId: string,
  outputDir: string = '/data',
  storageClient?: ZeroGStorageClient
): Promise<string> {
  let client = storageClient;
  if (!client) {
    // Create client from environment variables
    const rpcUrl = process.env.ZERO_G_RPC || 'https://evmrpc.0g.ai';
    const storageAddress = process.env.ZERO_G_STORAGE_CONTRACT;
    const privateKey = process.env.ZERO_G_PRIVATE_KEY;

    if (!storageAddress) {
      throw new Error('ZERO_G_STORAGE_CONTRACT environment variable required');
    }

    // Generate or load encryption key
    const encryptionPassword = process.env.INFT_ENCRYPTION_PASSWORD;
    let encryptionKey: Buffer | undefined;
    if (encryptionPassword) {
      encryptionKey = generateEncryptionKey(encryptionPassword);
    }

    client = new ZeroGStorageClient({
      rpcUrl,
      storageContractAddress: storageAddress,
      privateKey,
      encryptionKey,
    });
  }

  return client.loadFrom0gStorage(inftId, outputDir);
}
